lib/AST/Decl.cpp

```diff
@@ -610,37 +610,58 @@ case DeclKind::ID: return cast<ID##Decl>(this)->getLocFromSource();
   llvm_unreachable("Unknown decl kind");
 }
 
-const Decl::CachedExternalSourceLocs *Decl::getSerializedLocs() const {
-  if (CachedSerializedLocs) {
-    return CachedSerializedLocs;
-  }
-  auto *File = cast<FileUnit>(getDeclContext()->getModuleScopeContext());
-  assert(File->getKind() == FileUnitKind::SerializedAST &&
-         "getSerializedLocs() should only be called on decls in "
-         "a 'SerializedASTFile'");
-  auto Locs = File->getBasicLocsForDecl(this);
-  if (!Locs.hasValue()) {
-    static const Decl::CachedExternalSourceLocs NullLocs{};
-    return &NullLocs;
-  }
-  auto *Result = getASTContext().Allocate<Decl::CachedExternalSourceLocs>();
-  auto &SM = getASTContext().SourceMgr;
-#define CASE(X)                                                               \
-Result->X = SM.getLocFromExternalSource(Locs->SourceFilePath, Locs->X.Line,   \
-                                       Locs->X.Column);
-  CASE(Loc)
-  CASE(StartLoc)
-  CASE(EndLoc)
-#undef CASE
+const ExternalDeclLocs &Decl::getSerializedLocs(bool Resolve) const {
+  auto &Context = getASTContext();
+  ExternalDeclLocs *Result = nullptr;
+  if (auto EL = Context.getExternalLocs(this)) {
+    Result = *EL;
+  } else {
+    static ExternalDeclLocs NullLocs{};
+    auto *File = cast<FileUnit>(getDeclContext()->getModuleScopeContext());
+    if (File->getKind() != FileUnitKind::SerializedAST)
+      return NullLocs;
+
+    auto Positions = File->getBasicPositionsForDecl(this);
+    if (!Positions.hasValue()) {
+      Result = &NullLocs;
+    } else {
+      Result = getASTContext().Allocate<ExternalDeclLocs>();
+      Result->Positions = getASTContext().AllocateObjectCopy(*Positions);
+    }
+    Context.setExternalLocs(this, Result);
+  }
+
+  // Not all callers need SourceLoc's. Creating them requires loading the file,
+  // so avoid that where we can.
+  if (Result->Positions && Resolve && !Result->Resolved) {
+    auto &SM = getASTContext().SourceMgr;
+    auto *Positions = Result->Positions;
+
+    auto ResolveLoc = [&](const SourcePosition &Pos) -> SourceLoc {
+      // If the decl had a presumed loc, create its virtual file so that
+      // getPresumedLineAndColForLoc works from serialized locations as well.
+      if (Pos.Directive.isValid()) {
+        auto &SL = Pos.Directive;
+        SourceLoc Loc = SM.getLocForOffset(Result->BufferID, SL.Offset);
+        SM.createVirtualFile(Loc, SL.Name, SL.LineOffset, SL.Length);
+      }
+      return SM.getLocForOffset(Result->BufferID, Pos.Offset);
```

Comment:
I was thinking about this more while making some changes here...

There's no guarantee that the underlying file hasn't changed, so this could currently assert (with assertions enabled) or end up crashing <somewhere> (when disabled). I'm not sure what the best thing to do here is. We can either:
  1. Use `SM.getLocForBufferStart(Result->BufferID).getAdvancedLocOrInvalid(...)` which will give an invalid SourceLoc if it's out of range, or
  2. Do a similar check to the one in SwiftSourceDocInfo, ie. check that the file hasn't been modified. In this case there would still be the question as to whether we give invalid SourceLoc's or just return None.

I definitely need to add some tests for these cases as well.

Reply:
> Do a similar check to the one in SwiftSourceDocInfo, ie. check that the file hasn't been modified

This seems reasonable, particularly if we can do it only once for a particular file.  It's better to get no location than a wrong location.

> In this case there would still be the question as to whether we give invalid SourceLoc's or just return None.

Either seems fine in the abstract, as long as we cache the failure and won't try to deserialize it again.

Reply:
That's a good point re. once per file. Could keep a buffer id -> last modified/stat in source manager?